sycl/doc/SYCLRuntimeRFC.md

```diff
@@ -0,0 +1,97 @@
+      }
+    });
+  });
+}
+```
+
+The lambda passed to the `sycl::handler::single_task` function
+represents a device kernel. When a SYCL application is started or a
+dynamic SYCL library is loaded, all device images contained in the
+multi-targeted binary are registered in the program manager, one of the
+larger components of the SYCL runtime, using SYCL's `__tgt_register_lib`
+counterpart, `__sycl_register_lib`. The integration header, which is
+generated during device compilation and pre-included during host
+compilation, helps the runtime map SYCL kernel invocations to their
+corresponding symbols in the device code. The program manager uses the
+information contained in device image wrappers to map the kernel name to
+a set of device images, chooses an image compatible with the requested
+device, passes it to the backend to be just-in-time compiled if needed,
+and caches the result for reuse.
+
+The integration header also provides information about kernel arguments.
+It assumes that the lambda layout is the same for both device and host
+compilation, and the runtime uses the information about argument size
+and offsets to extract their values from the lambda object passed to
+`sycl::handler::single_task` and set them when enqueuing the kernel. The
+integration header also includes information about argument types, which
+is needed to handle special cases like `sycl::accessor` objects. For
+more details about the integration header, see (TODO add link to the
+header/footer RFC).
+
+## Scheduler
+
+Another large runtime component is the scheduler. The scheduler
+maintains a directed acyclic graph representation of the submitted
+command groups, managing dependencies and data movement required by the
+application code explicitly or with the use of accessors. Most of the
+actual synchronization is handled by the underlying device backend, with
+the scheduler simply enqueuing tasks as they are submitted and passing a
+list of deduced dependencies. There are some exceptions to this: a SYCL
+application may require access to some data on the host or submit an
+asynchronous host task. Such cases are handled by the scheduler directly
+by delaying the submission of device workloads until their host
+dependencies are satisfied.
+
+## Device headers and libraries
+
+SYCL runtime headers provide various algorithms and built-in functions
+to be used in device code. A large portion of these are simple enough to
+be implemented in the headers (for example, those directly lowered into
+SPIR-V built-ins), but some have more complex implementations (for
+example, fallback implementations for functionality that might not be
+supported natively). Those are available in the form of several
+pre-built device libraries that are linked with the user device code
+either during ahead-of-time compilation or at runtime by the program
+manager, using the information about device library dependencies
+embedded in device image wrappers by SYCL compilation tools.
```

Comment:
`libdevice` duplicates the functionality of libc and libm. I expect the AMD folks will ask us to use libc/libm instead of cotributing `libdevice` in its current form.

@jinge90, have you looked at libm/libc implementation? If yes, do you have a plan to integrate it into our implementation?

Reply:
There was a presentation about using libm from llvm.  The results are mixed results when comparing the routines from libm  with vendor provided implementation.

Reply:
During GPU offloading workshop we talked about having an option to customize libm implementation based on accuracy requirements. I expect this option to enable the use of SPIR-V math functions for libm implementation on Intel GPU.
AFAIK, @andykaylor is doing some work in this area.

Reply:
Hi, @bader 
I don't have detail plan to integrate llvm libc/libm into our implementation for now but I agree that we should align with LLVM libc/libm for libdevice in sycl. I have looked at libm/libc before and most generic versions of functions can be directly used in sycl such as <string.h>, <math.h>.
For the first step, maybe we can start from the LLVM libc/libm building system to support to use sycl toolchain to build required functions.
Thanks very much.

Reply:
@bader @jinge90 Do you think it makes sense to mention that we're looking in this direction in this RFC then? If so, how would you have it worded?

Reply:
@sergey-semenov not clear if Unified Runtime is sycl specific plugin or general UR which OpenMP is UR_plugin will  interface.  If it is  a plugin better to mention since we use similar  terms in OpenMP

Reply:
I think so. It seems there is not much we can say at this point, except that "some device libraries functionality overlaps with libc/libm implementation for GPU. We are looking for ways to re-use LLVM code for SYCL and welcome any feedback".